File: Cplusplus/238.product-of-array-except-self.cpp
/*
 * @lc app=leetcode.cn id=238 lang=cpp
 *
 * [238] 除自身以外数组的乘积
 */

// @lc code=start
class Solution {
public:
    // solution1 :
    //      use arry suffix product and prefix product
    //      O(3*n) , but need 2*n additional space
    // solution2 :
    //      only use prefix product
    vector<int> productExceptSelf(vector<int>& nums) {
        int Len = nums.size();
        vector<int> prefixProduct(Len, 1);
        if (Len == 0) return prefixProduct;

        // calculate prefix
        for(int idx=1; idx < Len; idx++) {
            prefixProduct[idx] = prefixProduct[idx-1] * nums[idx-1];
        }
        // calculate nums
        int suffix = 1;
        for (int idx = Len-2; idx >= 0; idx--) {
            suffix = suffix * nums[idx + 1];
            prefixProduct[idx] = prefixProduct[idx] * suffix;
        }
        return prefixProduct;
    }
};
// @lc code=end

File: Golang/238.product-of-array-except-self.go
package main

/*
*
Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

You must write an algorithm that runs in O(n) time and without using the division operation.
*/
func productExceptSelf(nums []int) []int {
	numsLen := len(nums)
	prefixProduct := make([]int, numsLen)

	prefixProduct[0] = 1
	for i := 1; i < numsLen; i++ {
		prefixProduct[i] = prefixProduct[i-1] * nums[i-1]
	}

	suffix := nums[numsLen-1]
	for idx := numsLen - 2; idx >= 0; idx-- {
		prefixProduct[idx] = prefixProduct[idx] * suffix
		suffix = suffix * nums[idx]
	}
	return prefixProduct
}
